### PR TITLE
Switch from flatgeobuffer to geomedea for GTFS

### PR DESCRIPTION
**Pull request by dabreegster**
CC @michaelkirk, I'm trying out geomedea for the use case I described in Discord!

| Metric  | flatgeobuffer | geomedea |
| --------- | ----------------- | -------------- |
| File size | 99MB         | 53MB |
| Bristol | 3.6MB in 23 requests | 5MB in 20 requests |
| Elephant & Castle | 6.4MB in 935 requests, 1.76 minutes | 9.4MB in 24 requests, 8.3s |

Bristol doesn't have many GTFS trip shapes intersecting the area, while E&C in London has loads.

Unless I'm measuring something wrong, the current approach with geomedea incurs more bandwidth, but through _way_ less requests and latency.

### CODE REVIEW

File: backend/src/gtfs/gmd.rs [hunk 1]
```diff
@@ -1,74 +1,80 @@
...
+            // TODO geozero to this translation better?
+            let geom = Geometry::LineString(geomedea::LineString::new(
+                linestring
+                    .0
+                    .into_iter()
+                    .map(|pt| geomedea::LngLat::degrees(pt.x, pt.y))
+                    .collect(),
+            ));
+            let mut props = Properties::empty();
+            // TODO bincode or something else?
```
Comment (dabreegster) [inline backend/src/gtfs/gmd.rs:33]: The size difference is probably coming from here, I need to rethink how to encode this. Most of the size comes from a bunch of `chrono::NaiveTime`s right now, which get encoded in a pretty naive way

Comment (michaelkirk) [reply backend/src/gtfs/gmd.rs:33]: Is `variant` some pre-encoding of all your properties into a single byte stream?

Internally geomedea is using bincode for encoding property and geometry, and then each page is zstd compressed.

Varints probably make sense for property data, but probably not currently for geometry data, until/unless I also implement delta encoding for geometries. I'd like to implement delta encoding but it'll require reworking some API internals.

File: backend/Cargo.toml [hunk 2]
```diff
@@ -34,10 +35,9 @@ wasm-bindgen = "0.2.87"
 wasm-bindgen-futures = "0.4.42"
 web-sys = { version = "0.3.64", features = ["console"] }
 web-time = "1.1.0"
-miniz_oxide = "0.7.4"
-itertools = "0.13.0"
 
 [target.'cfg(not(target_arch = "wasm32"))'.dependencies]
+geomedea = { git = "https://github.com/michaelkirk/geomedea" }
```
Comment (michaelkirk) [inline backend/Cargo.toml:40]: btw this should compile in wasm now.

**Comment by michaelkirk** [comment]
> Unless I'm measuring something wrong, the current approach with geomedea incurs more bandwidth, but through way less requests and latency.

It's not entirely surprising that geomedea might request more data. 

In FGB, there is a single buffer of uncompressed features. In FGB, since there is no compression, the index tells us exactly where each feature is in the file. Using this I implemented smart feature batching, so feature requests will only merge adjacent features into a single request if they are "close enough". 

To take advantage of compression, geomedea groups features into pages, so you have to download an entire page even if you only need one feature in the page. Because geomedea's features are in compressed pages, it's a little different. It can still be done, but I guess it'd be "page batches" rather than "feature batches". I haven't implemented this yet, but it should be doable in a non-breaking way.

**Comment by michaelkirk** [comment]
Could you do me a favor? 

RUST_LOG=debug

And give me the lines matching: `Finished using an HTTP client. used_bytes`

e.g.:
```
Finished using an HTTP client. used_bytes=839712, wasted_bytes=293690, req_count=4
Finished using an HTTP client. used_bytes=17, wasted_bytes=0, req_count=1
```

wasted_bytes should correspond to the bytes that could be gained by having more clever page-batching.

**Comment by michaelkirk** [comment]
> wasted_bytes should correspond to the bytes that could be gained by having more clever page-batching.

I had a go at "more clever page-batching" here:
https://github.com/michaelkirk/geomedea/pull/12

**Comment by michaelkirk** [comment]
I was looking at the network traffic for your existing FGB integration - and I feel like there must be a bug. It makes no sense for all those small nearby requests (4 bytes?!).

Screenshot 2024-07-18 at 17 01 39

I'm looking into that now.

File: backend/Cargo.toml (hunk 2 above)
Comment (dabreegster) [reply backend/Cargo.toml:40]: In WASM, I want to disable the `writer` feature, but otherwise enable it. I fiddled around with specifying features based on architecture and landed here as something that works, but I'll take another look if there's a way to be more clear here...

File: backend/src/gtfs/gmd.rs (hunk 1 above)
Comment (dabreegster) [reply backend/src/gtfs/gmd.rs:33]: I'm using `serde_json::to_vec` on ```
struct RouteVariant {
    // Per stop, (original ID and name)
    pub stop_info: Vec<(orig_ids::StopID, String)>,

    // Each one has an arrival time per stop
    pub trips: Vec<Vec<NaiveTime>>,

    // Metadata
    pub route: Route,
}
```
The space is dominated by the times. Dropping some precision and some delta encoding would make tons of sense there.

I'll also play with using `PropertyValue::Vec` of some integers manually, instead of this.

Comment (dabreegster) [reply backend/src/gtfs/gmd.rs:33]: https://docs.rs/chrono/latest/src/chrono/naive/time/serde.rs.html#5

OK, NaiveTime gets serde-ified as a string right now, that's amazing. Switching to something more appropriate...

**Comment by dabreegster** [comment]
With the new property encoding...

Elephant reads 6.3MB over 23 requests.
Finished using an HTTP client. used_bytes=156716, wasted_bytes=0, req_count=5
Finished using an HTTP client. used_bytes=3776001, wasted_bytes=1490472, req_count=17
Finished using an HTTP client. used_bytes=17, wasted_bytes=0, req_count=1

Bristol reads 2.5MB over 25 requests
Finished using an HTTP client. used_bytes=144956, wasted_bytes=1344, req_count=9
Finished using an HTTP client. used_bytes=1078601, wasted_bytes=473493, req_count=15
Finished using an HTTP client. used_bytes=17, wasted_bytes=0, req_count=1

So the new encoding is not giving that huge of an advantage, but still opens the way to doing something nicer later with delta encoding.

I'm going to merge this in now and continue to play with encoding / perf later on. It's a huge improvement with low work, so thanks so much for the new format, adding WASM support, and these page batching fixes!

**Comment by michaelkirk** [comment]
Here's Elephant & Castle with https://github.com/flatgeobuf/flatgeobuf/pull/376
Screenshot 2024-07-23 at 15 07 52

tldr; there was a bad bug in the http fetch implementation. It hadn't came up in the shape of my data and requests, so thanks for helping to uncover it.

With the bug fix, the two formats seem to be in the same ballpark of network transfer for your queries.
